src/engine/SCons/Job.py

```diff
@@ -375,55 +375,83 @@ def start(self):
-                    task = self.taskmaster.next_task()
-                    if task is None:
-                        break
 
-                    try:
-                        # prepare task for execution
-                        task.prepare()
-                    except:
-                        task.exception_set()
-                        task.failed()
-                        task.postprocess()
-                    else:
-                        if task.needs_execute():
-                            # dispatch task
-                            self.tp.put(task)
-                            jobs = jobs + 1
-                        else:
-                            task.executed()
-                            task.postprocess()
-
-                if not task and not jobs: break
-
-                # Let any/all completed tasks finish up before we go
-                # back and put the next batch of tasks on the queue.
-                while True:
-                    task, ok = self.tp.get()
-                    jobs = jobs - 1
-
-                    if ok:
+               # Process jobs until we find one that requires
+               # execution, and enqueue it.
+               while True:
+                  task = self.taskmaster.next_task()
+                  if not task:
+                     break
+
+                  try:
+                     # prepare task for execution
+                     task.prepare()
+                  except:
+                     task.exception_set()
+                     task.failed()
+                     task.postprocess()
+                  else:
+                     if task.needs_execute():
+                        # dispatch task
+                        self.tp.put(task)
+                        jobs = jobs + 1
+                        break
+                     else:
                         task.executed()
-                    else:
-                        if self.interrupted():
-                            try:
-                                raise SCons.Errors.BuildError(
-                                    task.targets[0], errstr=interrupt_msg)
-                            except:
-                                task.exception_set()
+                        task.postprocess()
+
+
+               # If 'task' is true-ish, then we enqueued a new
+               # task. If 'task' is falseish, then we didn't. If we
+               # didn't enqueue a task and we don't have any jobs
+               # pending, then there is nothing left to do.
+               if not (task or jobs):
+                  break
+
+
+               # If we are at maxjobs, then force a wait for a job
+               # completion so that we don't go over the limit by
+               # looping back up to next_task. If we didn't find any
+               # new tasks above, we must also block so that we don't
+               # busy loop.
+               should_block = False
+               should_block |= (jobs == self.maxjobs)
```

Comment:
As I mentioned on my proposed #3404 , I think we should be more aggressive than this. I would like for us to not block until task is None. Any time blocking on the main Python thread is wasted time where we could be gathering more tasks for dispatch.

Reply:
@grossag - This only blocks if we were either unable to find any new tasks to run, or if we are not allowed to issue any more jobs. In the former case, we won't be able to find any new tasks to run until we retire some jobs that we have issued, and processed their results unblocking access to new parts of the DAG. So I think it does what you are suggesting.

Reply:
But this also blocks if there are tasks available to run but we are not able to issue any more jobs. A more efficient scheduler would actually keep calling taskmaster.next_task() if jobs==self.maxjobs and put them in a list. This is because next_task() is not instantaneous in that it generates scan requests, which goes to disk. Any time spent blocking while next_task() would return more tasks is time wasted.

Reply:
Here is a scheduler I wrote as part of remote caching support. Pull request is coming eventually, but I need to improve testability. This scheduler will focus on doing three things as efficiently as possible: (1) gather tasks, (2) fetch tasks from remote cache, and (3) deploy jobs to the thread pool. It will only wait on jobs results if there are no tasks left to gather and no fetches pending to the remote cache. Here is the code: https://gist.github.com/grossag/9e7ef2e8c3c02584152c67812af7cd3a .

I post this because it demonstrates what I mean about not waiting on jobs unless taskmaster.next_task() returns None.

Reply:
@grossag - OK, yes, I see what you are saying. However, this is a meaningful improvement over the current behavior, which exhibits unfortunate sawtooth behavior. Would you be OK with seeing this merge now, even if it is later supplanted by your even better approach? Keeping things iterative, etc.

Reply:
@grossag - Actually, having thought about this a little more, I think it is easy to apply that idea to this implementation. I'll give it a try.